distribution/xfer/download.go

```diff
@@ -62,6 +63,9 @@ type DownloadDescriptor interface {
 	// if it is unknown (for example, if it has not been downloaded
 	// before).
 	DiffID() (layer.DiffID, error)
+	// ForeignSource returns nil or the foreign layer's information and
+	// source.
+	ForeignSource() *distribution.Descriptor
```

Comment:
I've thought about this a bit previously, but the issue is that distribution metadata only tracks layers that are pulled or push. We need to preserve the foreign source information for layers brought in with `docker load` as well. It's technically possible to use the distribution metadata, but it would leak this metadata out of the context of push/pull.